pkg/controller/projects/controller.go

```diff
@@ -0,0 +1,586 @@
+}
+
+type connector struct {
+	kube              client.Client
+	newArgocdClientFn func(clientOpts *apiclient.ClientOptions) project.ProjectServiceClient
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return nil, errors.New(errNotProject)
+	}
+	cfg, err := clients.GetConfig(ctx, c.kube, cr)
+	if err != nil {
+		return nil, err
+	}
+	return &external{kube: c.kube, client: c.newArgocdClientFn(cfg)}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client projects.ProjectServiceClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotProject)
+	}
+
+	if meta.GetExternalName(cr) == "" {
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	projectQuery := project.ProjectQuery{
+		Name: meta.GetExternalName(cr),
+	}
+
+	project, err := e.client.Get(ctx, &projectQuery)
+	if projects.IsErrorProjectNotFound(err) {
+		return managed.ExternalObservation{}, nil
+	}
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetFailed)
+	}
+
+	current := cr.Spec.ForProvider.DeepCopy()
+	lateInitializeProject(&cr.Spec.ForProvider, &project.Spec)
+
+	cr.Status.AtProvider = generateProjectObservation(project)
+	cr.Status.SetConditions(xpv1.Available())
+
+	return managed.ExternalObservation{
+		ResourceExists:          true,
+		ResourceUpToDate:        isProjectUpToDate(&cr.Spec.ForProvider, project),
+		ResourceLateInitialized: !cmp.Equal(current, &cr.Spec.ForProvider),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotProject)
+	}
+
+	projCreateRequest := generateCreateProjectOptions(cr)
+
+	resp, err := e.client.Create(ctx, projCreateRequest)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreateFailed)
+	}
+
+	meta.SetExternalName(cr, resp.Name)
+
+	return managed.ExternalCreation{
+		ExternalNameAssigned: true,
+	}, errors.Wrap(nil, errKubeUpdateFailed)
+}
+
+func (e *external) Update(ctx context.Context, mg resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mg.(*v1alpha1.Project)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errNotProject)
+	}
+	projQuery := project.ProjectQuery{
+		Name: meta.GetExternalName(cr),
+	}
+
+	proj, err := e.client.Get(ctx, &projQuery)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdateFailed)
+	}
+
+	projUpdateRequest := generateUpdateProjectOptions(cr, proj)
+
+	_, err = e.client.Update(ctx, projUpdateRequest)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(err, errUpdateFailed)
+	}
+
+	return managed.ExternalUpdate{}, nil
```

Comment:
Those last lines can be merged into `return managed.ExternalUpdate{}, errors.Wrap(err, errUpdateFailed)` since `errors.Wrap` will return `nil` in case `err` is `nil`.